catkin_ws/src/state_estimation/src/sensors.py

```diff
@@ -132,60 +141,51 @@ def __init__(self, imu):
         self.auv_dvl_offset = np.array([auv_dvl_offset_x, auv_dvl_offset_y, auv_dvl_offset_z])
 
         self.imu = imu
+        self.imu_front_camera = imu_front_camera
         self.q_dvlref_nwu = None
         
-        rospy.Subscriber("/sensors/dvl/pose", DeadReckonReport, self.dr_cb)
+        rospy.Subscriber("/sensors/dvl/status", Int32, self.sensor_status_cb)
+        rospy.Subscriber("/sensors/dvl/pose", DeadReckonReport, self.dead_reckon_cb)
 
-    def dr_cb(self, dr_msg):
+    def dead_reckon_cb(self, msg):
         # quaternion/position of dvl relative to dvlref 
-        # q_dvlref_dvl = #quaternion.from_euler_angles(dr_msg.roll*RAD_PER_DEG, dr_msg.pitch*RAD_PER_DEG, dr_msg.yaw*RAD_PER_DEG)
-        q_dvlref_dvl = transformations.quaternion_from_euler(dr_msg.roll*RAD_PER_DEG, dr_msg.pitch*RAD_PER_DEG, dr_msg.yaw*RAD_PER_DEG)
+        # q_dvlref_dvl = quaternion.from_euler_angles(msg.roll * RAD_PER_DEG, msg.pitch * RAD_PER_DEG, msg.yaw * RAD_PER_DEG)
+        q_dvlref_dvl = transformations.quaternion_from_euler(msg.roll * RAD_PER_DEG, msg.pitch * RAD_PER_DEG, msg.yaw * RAD_PER_DEG)
         q_dvlref_dvl = np.quaternion(q_dvlref_dvl[3], q_dvlref_dvl[0], q_dvlref_dvl[1], q_dvlref_dvl[2])
         q_dvlref_auv = q_dvlref_dvl * self.q_dvl_auv
-        pos_dvlref_dvl = np.array([dr_msg.x, dr_msg.y, dr_msg.z])
+        pos_dvlref_dvl = np.array([msg.x, msg.y, msg.z])
+
+        is_imu_active = self.imu.get_is_active()
+        is_imu_front_camera_active = self.imu_front_camera.get_is_active()
+
+        imu_q_nwu_auv = self.imu.q_nwu_auv
+        imu_front_camera_q_nwu_auv = self.imu_front_camera.q_nwu_auv
 
-        #update dvl ref frame using imu
-        if self.imu.isActive():
+        # update dvl ref frame using imu
+        if is_imu_active:
             q_nwu_auv = self.imu.q_nwu_auv
             self.q_dvlref_nwu = q_dvlref_auv * q_nwu_auv.conjugate() 
+        elif is_imu_front_camera_active:
+            q_nwu_auv = self.imu_front_camera.q_nwu_auv
+            self.q_dvlref_nwu = q_dvlref_auv * q_nwu_auv.conjugate() 
 
         if self.q_dvlref_nwu is None: return
 
         temp_pos_auv = self.q_dvlref_nwu.conjugate() * np.quaternion(0, pos_dvlref_dvl[0], pos_dvlref_dvl[1],pos_dvlref_dvl[2]) * self.q_dvlref_nwu
         pos_auv = np.array([temp_pos_auv.x, temp_pos_auv.y, temp_pos_auv.z])
-        temp_dvl_auv_offset_rotated = self.imu.q_nwu_auv * np.quaternion(0,self.auv_dvl_offset[0],self.auv_dvl_offset[1], self.auv_dvl_offset[2]) * self.imu.q_nwu_auv.conjugate()
+        if is_imu_active or not is_imu_front_camera_active:
+            temp_dvl_auv_offset_rotated = imu_q_nwu_auv * np.quaternion(0,self.auv_dvl_offset[0],self.auv_dvl_offset[1], self.auv_dvl_offset[2]) * imu_q_nwu_auv.conjugate()
+        else:
+            temp_dvl_auv_offset_rotated = imu_front_camera_q_nwu_auv * np.quaternion(0,self.auv_dvl_offset[0],self.auv_dvl_offset[1], self.auv_dvl_offset[2]) * imu_front_camera_q_nwu_auv.conjugate()
```

Comment:
just adding some if's to decide if and which IMU to use

Reply:
tbh i would get rid of this. if the imu turns off id rather the auv just stop then do some unpredicatable behaviour

Reply:
I don't like the ifs either but for a different reason, since we already have a tree of sensors and their priorities in state aggregator I would pass that into DVL (pass a callable into DVL or just setter for the current IMU or something)

Reply:
its fair that if IMU dies DVL should die. Regardless tho overall DVL class is gonna change a ton so i dont really care tbh